test/compiler/irpasses.jl

```diff
@@ -1421,3 +1421,20 @@ let src = code_typed1(sroaunswitchunionstruct1, Tuple{Bool, Int, Float64})
 end
 @test sroaunswitchunionstruct2(true, 1, 1.0) === 1
 @test sroaunswitchunionstruct2(false, 1, 1.0) === 1.0
+
+# Test SROA of union into getfield
+struct SingleFieldStruct1
+	x::Int
+end
+struct SingleFieldStruct2
+	x::Int
+end
+function foo(b, x)
+	if b
+		f = SingleFieldStruct1(x)
+	else
+		f = SingleFieldStruct2(x)
+	end
+	getfield(f, :x) + 1
+end
+@test foo(true, 1) == 2
```

Comment:
Don’t we also need to test that there are no getfield calls?

Reply:
This PR tries to fix cases where SROA pass accidentally eliminates `getfield` call too aggressively. So we actually need to preserve the `getfield` call for this case.

Reply:
Well, we could enhance sroa to handle this case, but that's an entirely separate change. This just checks correctness.